### ISSUE TITLE
Could there be a more forgiving mode re: missing plugins?

### ISSUE THREAD

**Issue by beverloo** [mdx-editor/editor#95]
First off - thank you for publishing this library, I'm a happy user 😃 

I'd like to ask whether you've considered a more forgiving mode, as it's fairly straightforward for my users to (accidentally) break the editor. Hypothetically, say that I _don't_ load the headings plugin (#69) but the Markdown content still contains a heading, they're now faced with a JavaScript error that means nothing to them.

I feel like this is almost requiring me to load all the plugins—even though that would still leave edge cases that could cause the editor to break when using JSX elements, admonitions and so on, or choose not to expose the Markdown at all. An alternative could be to just render the invalid (due to a missing plugin) Markdown as if it were text.

**Comment by petyosi** [comment]
That's a good discussion to have. To start off, I do believe that most common setups should include the standard markdown plugins, like headings, lists, etc. In my mind, there are only two reasons why you might *not* want to include some of them:

* You're building a restrictive input field - like the Slack message input, or a comment field, so you don't want to let users insert some formatting that breaks your logic.
* You need custom logic for the node type.

The JSX, Directive, and Code Block plugins are more complex of course, so including them means that you're building something for a trusted set of users who know what they are doing. In general, I designed the source mode for those environments, otherwise it's quite dangerous indeed.

This being said, a catch-all behavior is nice one, but I am not sure I can do something that will work for everyone. However, the current plugin mechanism lets you (relatively) easy to implement such behavior. [Take a look at this snippet](https://codesandbox.io/s/catch-all-plugin-c7d62v?file=/src/App.tsx), for example. It handles any unrecognized node and inserts its contents in a plain text mode. The code blocks and the directives have similar mechanisms as well. 

Let me know what you think about this. Happy to continue this conversation, chances are there's some improvement worth putting in as a built-in behavior.

**Comment by beverloo** [comment]
Thank you for the thorough response!

I agree that most plugins should be loaded most of the time—supporting "regular" Markdown / MDX as a minimum baseline is sensible. I used the headings example because there was an issue about it, it's the more advanced markup types that I'm interested in.

The example that I ran into last week was me mistyping a JSX element (`<Remotecontent />` rather than `<RemoteContent />`) which I could only correct directly in the database.

The example that I ran into today was that a user used [text expressions](https://mdxjs.com/docs/what-is-mdx/#expressions) incorrectly by accident (`Hello, {firstName}!`), which led to an exception:
```
Error: no MdastImportVisitor found for mdxTextExpression {"type":"mdxTextExpression","value":"firstName","position":{"start":
```

In other words, and to be concrete about my goal: I don't want my users to be able to break the editor, with the only available recourse being a direct database edit. That means that I'd like to have a working editor no matter what ends up being passed in `<MDXEditor markdown="# foo" />`.

Thank you for sharing the `MdastImportVisitor` snippet! The example given doesn't solve the problem, but the system seems powerful so I'll try to come up with something that does :)

**Comment by beverloo** [comment]
Also, to be clear, it's perfectly fine if that means that MDXEditor isn't what I should be using. You should follow your own design principles!

**Comment by petyosi** [comment]
Thank you for extending the context further, this is very helpful. That 100% makes sense and, in fact, aligns with my general goals with the editor. I do want to make it work **effectively** with the most amount of markdown you can encounter in the wild, while also allowing the markdown to be edited manually and not ending up in a hard to recover state. It's a matter of time and work to get there, there's no general design mismatch. In fact, it's not complicated from a technical standpoint, but more a matter of thinking it through from the UX perspective.

Handling unrecognized JSX elements in a graceful way is necessary. I haven't encountered the JSX expressions yet, but I guess they can be treated as some variable fields where the users can edit them in a nested input. 

I will keep this issue open, as I feel like it is likely to attract some good input, and I will post updates as I make some progress related to the matter. Needless to say, I'm more than happy to accept contributions as well.

**Comment by live2ride** [comment]
I would like to say thank you for the editor and the snippet!

However in my opinion editor default state should never break due to input, so some version of catchAllPlugin should be the default last plugin.

perhaps something like this

```jsx

const catchAllVisitor: MdastImportVisitor<any> = {
  testNode: () => true,
  visitNode: ({ mdastNode, actions, lexicalParent }) => {
    const paragraph = $createParagraphNode();
    paragraph.append($createTextNode(mdastNode?.value || ""));
    if(process.env.NODE_ENV !== "production"){
      console.warn(`unhandled node reverted to paragraph, type: ${mdastNode?.type}, value: `, mdastNode?.value )
    }
    (lexicalParent as ElementNode).append(paragraph);
  }
};

const [catchAllPlugin] = realmPlugin({
  id: "catchAll",
  systemSpec: system(() => ({})),
  init: (realm) => {
    realm.pubKey("addImportVisitor", catchAllVisitor);
  }
});

MDXEditor.tsx
export const MDXEditor = React.forwardRef<MDXEditorMethods, MDXEditorProps>((props, ref) => {
  return (
    <RealmPluginInitializer
      plugins={[
        ...(props.plugins || []),
        catchAllPlugin()
      ]}
    >
      xxxx
    </RealmPluginInitializer>
  )
```

until then, everyone else can use below:
```jsx

<MDXEditor  plugins={[...ALL_PLUGINS, catchAllPlugin()]} />

```

**Comment by petyosi** [comment]
I'm with you here, considering of a proper way to preserve the unrecognized html tags somehow, so that 

1) the editor does not break
2) the markdown does not get damaged after editing

**Comment by petyosi** [comment]
I've done a few things that make it easier to edit unknown markdown in the latest release. The editor exposes an `onError` callback prop that you can use to handle unsupported content. In addition to that, the diff/source mode can now be used to fix those.

**Comment by rob-at-airwalk** [comment]
hey @petyosi , these fixes look great for normal markdown, but it doesn't seem to handle MDX type content unless I'm mistaken?
For example, if my markdown is something like
```
here is some demo content
<Alert>this component doesn't exist</Alert>
```
I want to be able to ignore the Alert component (if it doesn't exist), but not change the markdown. the user should be able to edit in source mode without the MDX being removed. so exactly the same as it works for normal markdown where the plugin doesn't exist, but for MDX components.

maybe I'm missing something, but any pointers would be much appreciated.

doing something like this as suggested in another thread, does avoid an error, but also the markdown is modified:

```
const catchAllVisitor = {
  testNode: () => true,

  visitNode: ({ mdastNode, actions, lexicalParent }) => {
    const paragraph = $createParagraphNode();
    try {
      paragraph.append($createTextNode(mdastNode.children[0].value));
    } catch (err) {
      console.error("Editor:catchAllPlugin:error: ", err);
    }
    lexicalParent.append(paragraph);
  },
};
```

**Comment by petyosi** [comment]
@rob-at-airwalk what you describe is not handled well at the moment. To make that work, the jsxPlugin should support a fallback component descriptor of some sort - right now it only handles "known" components. 

If you have the capacity, take a look at the implementation there. Happy to accept a PR on the matters.

**Comment by yakkomajuri** [comment]
a lot of the imports outlined in the solution here don't seem to exist anymore. what's the current approach to having a catch all plugin? @petyosi
